metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/messages/CheckpointingMessage.scala

```diff
@@ -0,0 +1,31 @@
+package io.iohk.metronome.checkpointing.service.messages
+
+import io.iohk.metronome.checkpointing.models.Ledger
+import io.iohk.metronome.core.messages.{RPCMessage, RPCMessageCompanion}
+
+/** Checkpointing specific messages that the HotStuff service doesn't handle,
+  * which is the synchronisation of committed ledger state.
+  *
+  * These will be wrapped in an `ApplicationMessage`.
```

Comment:
Why so? I would expect the networking and RPC layers to be unrelated, and so their messages too.

Reply:
The `networking` module is indeed unrelated to the RPC layer, in the sense that it doesn't know what kind of messages it is transmitting, it's [generic in the message type](https://github.com/input-output-hk/metronome/blob/develop/metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala#L21). 

But in the RPL layer I see our application as having two components:
* generic HotStuff consensus, which can be applied to any `Agreement`
* agreement specific extras, which in our case is Checkpointing which can directly exchange the Ledger state, but it could be something else that may require its own different set of messages.

So I want the `RemoteConnectionManager` to be able to handle the union of these messages, hence the wrapping, and we'll have something like a `RemoteConnectionManager[Task, CheckpointingAgreement.PKey, Either[HotStuffMessage, CheckpointingMessage]`. I just made that explicit by having a `DuplexMessage` instead of the `Either`. 

Does that answer your question? `ApplictionMessage` has nothing to do with the network, it's just a wrapper that I came up with so that we can compose a network that has application side messages, _if_ if needs them.

So our network - RPC stack would look like this:
```
              -----------
              | Network |
              -----------
                  |
           -----------------
           |               |
     ------------     -----------------
     | HotStuff |     | Checkpointing |
     ------------     -----------------
           |
    -------------------
    |                 |
-------------      --------
| Consensus |      | Sync |
-------------      --------
```
With some further pipes between components that need to send messages to each other as well.

Reply:
The HotStuff side of that diagram is reflected in the [HotStuffService](https://github.com/input-output-hk/metronome/pull/25/files#diff-0f291caa9674678c06eec2f7fe2c17c89cb58bf117c8cf5a67ecd8c03d3360bb)

Reply:
I may be a bit confused here. Is `CheckpointMessage` meant be exchanged between Metronome nodes or between a Metronome node and an associated PoW node?

Reply:
I'm sorry, I found it difficult to come up with names; I named it after the eventual `CheckpointingService`, so `ConsensusMessage` goes to the `ConsensusService`, `SyncMessage` goes to the `SyncService`. 

The `CheckpointingMessage` is stuff that goes between Metronome nodes, that's why it's under the "Remote messaging protocol" PR. 

The proposal for messages that go between the Metronome node and the associated PoW node are in https://github.com/input-output-hk/metronome/pull/24

Reply:
The `CheckpointingMessage` is what allows two nodes to synchronise their ledger state; it could work in different ways, though, like synchronising a Merkle Patricia Trie, for just the parts that changed. It's just that in our case we know that our `Ledger` resets with every checkpoint, so it's small enough to be exchanged in a single message.